### PR TITLE
Refactor FrontEnds and update vertex4 

### PR DESCRIPTION
**Pull request by houpc**

### CODE REVIEW

File: src/computational_graph/abstractgraph.jl [hunk 1]
```diff
@@ -289,50 +289,45 @@ Base.:(==)(a::AbstractGraph, b::AbstractGraph) = Base.isequal(a, b)
...
-    #     !all(isequiv.(subgraphs(a)[pa], subgraphs(b)[pb], args...)) && return false
-    # else
-    a_pairs = collect(zip(subgraphs(a), subgraph_factors(a)))
-    b_pairs = collect(zip(subgraphs(b), subgraph_factors(b)))
+    # Compare groups of subgraphs with the same factor
+    a_pairs = zip(subgraphs(a), asubg_factors)
+    b_pairs = collect(zip(subgraphs(b), bsubg_factors))
+    # b_pairs = zip(subgraphs(b), subgraph_factors(b))
+    # matched = falses(length(asubg_factors))
     for (suba, suba_factor) in a_pairs
```
Comment (dcerkoney) [inline src/computational_graph/abstractgraph.jl:318]: I don't see any need for `idx` and collected `b_pairs` here, we just want to look for the first non-match and return false asap if we do. We have also already tested that subg(a) and subg(b) are same-length at this stage. How about the following?

```
a_pairs = zip(subgraphs(a), asubg_factors)
b_pairs = zip(subgraphs(b), bsubg_factors)
for (suba, suba_factor) in a_pairs
    subequiv = false
    for (subb, subb_factor) in b_pairs
        return suba_factor == subb_factor && isequiv(suba, subb, args...)
    end
end
return true  # base case: suba and subb are empty
```

File: src/frontend/parquet/parquet.jl [hunk 2]
```diff
@@ -202,12 +208,22 @@ end
...
 include("common.jl")
-
 include("operation.jl")
-
 include("filter.jl")
-include("vertex4.jl")
+include("to_graph.jl")
 
+const vertex4I_diags = Dict{Int,Vector{Graph}}()
+function initialize_vertex4I_diags(; filter=[NoHartree], spinPolarPara::Float64=0.0)
```
Comment (dcerkoney) [inline src/frontend/parquet/parquet.jl:216]: Can we add a docstring? Why is this function necessary / why only partitions (3, 0, 0) and (4, 0, 0)?

**Review by dcerkoney — APPROVED** [review]
Looks good to me, but I think we need to add some tests before merging.

File: src/computational_graph/abstractgraph.jl [hunk 3]
```diff
@@ -289,50 +289,45 @@ Base.:(==)(a::AbstractGraph, b::AbstractGraph) = Base.isequal(a, b)
...
         match_found = false
         for (idx, (subb, subb_factor)) in enumerate(b_pairs)
             if suba_factor == subb_factor && isequiv(suba, subb, args...)
+                # if !matched[idx] && suba_factor == subb_factor && isequiv(suba, subb, args...)
                 deleteat!(b_pairs, idx)
```
[the remark this answers is not in the document]
Comment (houpc) [reply src/computational_graph/abstractgraph.jl:323]: It is indeed inefficient now. I think we need to rewrite a 'local' equiv function later.

File: src/frontend/parquet/parquet.jl (hunk 2 above)
Comment (houpc) [reply src/frontend/parquet/parquet.jl:226]: This function initializes the fully-irreducible diagrams from GV, which will be used in Parquet.vertex4.

File: src/computational_graph/optimize.jl [hunk 4]
```diff
@@ -43,21 +36,23 @@ function optimize!_v0(graphs::Union{Tuple,AbstractVector{<:AbstractGraph}}; verb
 end
 
 """
-    function optimize(graphs::Union{Tuple,AbstractVector{<:AbstractGraph}}; verbose=0, normalize=nothing)
+    function optimize(graphs::Union{Tuple,AbstractVector{<:AbstractGraph}}; level=0, verbose=0, normalize=nothing)
 
-    Optimizes a copy of given `graphs`. Removes duplicated leaves, merges chains, and merges linear combinations.
+    Optimizes a copy of given `graphs`. Removes duplicated nodes (when `level > 0`) or leaves, flattens chains,
```
Comment (dcerkoney) [inline src/computational_graph/optimize.jl:41]: *Removes duplicated leaves, flattens operator chains, merges linear combinations, and removes zero-valued subgraphs. When `level > 0`, also removes duplicated intermediate nodes.

File: src/computational_graph/optimize.jl [hunk 5]
```diff
@@ -43,21 +36,23 @@ function optimize!_v0(graphs::Union{Tuple,AbstractVector{<:AbstractGraph}}; verb
...
-    function optimize(graphs::Union{Tuple,AbstractVector{<:AbstractGraph}}; verbose=0, normalize=nothing)
+    function optimize(graphs::Union{Tuple,AbstractVector{<:AbstractGraph}}; level=0, verbose=0, normalize=nothing)
 
-    Optimizes a copy of given `graphs`. Removes duplicated leaves, merges chains, and merges linear combinations.
+    Optimizes a copy of given `graphs`. Removes duplicated nodes (when `level > 0`) or leaves, flattens chains, 
+    merges linear combinations, and removing zero-valued subgraphs.
 
 # Arguments:
 - `graphs`: A tuple or vector of graphs.
+- `level`: Optimization level (default: 0). A value greater than 0 triggers more extensive but slower optimization processes, such as removing duplicated nodes.
```
Comment (dcerkoney) [inline src/computational_graph/optimize.jl:46]: *...duplicated intermediate nodes.

File: src/frontend/parquet/operation.jl [hunk 6]
```diff
@@ -139,3 +139,114 @@ end
...
 
+"""
+    update_extKT!(diags::Vector{Graph}, para::DiagPara, legK::Vector{Vector{Float64}})
+
+    Update the external momenta (`extK`) and external times (`extT`) of all the nodes in a vector of graphs in-place.
+
+# Arguments
+- `diags::Vector{Graph}`: A vector of `Graph` objects.
+- `para::DiagPara`: parameters reconstructed in the graphs. Its `firstTauIdx` will update the `extT` of graphs.
+- `legK::Vector{Vector{Float64}}`: basus of the external momenta for the legs of the diagram as [left in, left out, right in, right out].
```
Comment (dcerkoney) [inline src/frontend/parquet/operation.jl:150]: *basis

File: src/frontend/parquet/operation.jl [hunk 7]
```diff
@@ -139,3 +139,114 @@ end
...
+"""
+    update_extKT(diags::Vector{Graph}, para::DiagPara, legK::Vector{Vector{Float64}}) -> Vector{Graph}
+
+    Returns a new vector of graphs with updated external momenta (`extK`) and external times (`extT`), 
+    based on the provided graphs, parameters, and external legs' momenta.
+
+# Arguments
+- `diags::Vector{Graph}`: A vector of `Graph` objects.
+- `para::DiagPara`: parameters reconstructed in the graphs. Its `firstTauIdx` will update the `extT` of graphs.
+- `legK::Vector{Vector{Float64}}`: basus of the external momenta for the legs of the diagram as [left in, left out, right in, right out].
```
Comment (dcerkoney) [inline src/frontend/parquet/operation.jl:253]: *basis

File: src/frontend/parquet/operation.jl [hunk 8]
```diff
@@ -139,3 +139,114 @@ end
...
+    Returns a new vector of graphs with updated external momenta (`extK`) and external times (`extT`), 
+    based on the provided graphs, parameters, and external legs' momenta.
+
+# Arguments
+- `diags::Vector{Graph}`: A vector of `Graph` objects.
+- `para::DiagPara`: parameters reconstructed in the graphs. Its `firstTauIdx` will update the `extT` of graphs.
+- `legK::Vector{Vector{Float64}}`: basus of the external momenta for the legs of the diagram as [left in, left out, right in, right out]. 
+
+# Returns
+- `Vector{Graph}`: A new vector of `Graph` objects with updated `extK`, `extT`, and `para` (if existed) properties for each node.
```
Comment (dcerkoney) [inline src/frontend/parquet/operation.jl:256]: *if existing

**Review by dcerkoney — APPROVED** [review]
Looks good to me modulo a few docstring typos

**Review by kunyuan — APPROVED** [review]

**Pull request merged**